tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py

```diff
@@ -0,0 +1,385 @@
+            self.error = f"Unexpected error: {str(e)}"
+            print(f"[{self.ip}] ✗ Unexpected error: {str(e)}")
+
+
+def check_cluster_liveness(nodes: List[str], port: int) -> Optional[Dict]:
+    """
+    Goal: can we naiively compile a CFR command to forcibly recover a cluster
+
+    detail:
+        For each provided node, gather that node's perspective on every other node's liveness to check whether we can naiively apply CFR.
+        We can naiively perform CFR if we, the operator, can reach all living nodes, and all living nodes can reach eachother.
+
+        Any more-intersting situaions probably require a closer look.
+
+    detail detail:
+        consensus here refers to whether everyone agrees
+         - all nodes agree with eachother if every response we get has the same list of living nodes
+         - we agree with the nodes if the number of responses equals the number of living nodes
+         - both agreeing is consensus
+
+        If we have consensus, we can create the parameters for a CFR invocation
+
+    Returns a dict with:
+        - alive_nodes: set of node IDs that are alive (if consensus reached)
+        - dead_nodes: set of node IDs that are dead (if consensus reached)
+        - all_node_ids: set of all node IDs in the cluster
+        - responding_ips: list of IPs that responded successfully
+        - consensus: True if the we have full agreement per the detail note
+    """
+    print(f"\n{'=' * 70}")
+    print(f"Checking Cluster Liveness")
+    print(f"{'=' * 70}")
+    print(f"Querying {len(nodes)} nodes for broker status...")
+    print(f"{'=' * 70}\n")
+
+    # Create checker instances for each node
+    checkers = [LivenessChecker(ip, port) for ip in nodes]
+
+    # Create and start threads
+    threads = []
+    for checker in checkers:
+        thread = threading.Thread(target=checker.check)
+        threads.append(thread)
+        thread.start()
+
+    # Wait for all threads to complete
+    for thread in threads:
+        thread.join()
+
+    # Analyze results
+    print(f"\n{'=' * 70}")
+    print("Liveness Check Results")
+    print(f"{'=' * 70}")
+
+    successful_checkers = [c for c in checkers if c.alive_nodes is not None]
+    failed_checkers = [c for c in checkers if c.error is not None]
+
+    print(f"\nResponding nodes: {len(successful_checkers)}/{len(nodes)}")
+    print(f"Non-responding nodes: {len(failed_checkers)}/{len(nodes)}")
+
+    if not successful_checkers:
+        print("\n✗ ERROR: No nodes responded successfully. Cannot proceed.")
+        return None
+
+    # Compare alive node sets from all responding nodes
+    alive_node_sets = [c.alive_nodes for c in successful_checkers]
+    first_alive_set = alive_node_sets[0]
+    all_agree = all(alive_set == first_alive_set for alive_set in alive_node_sets)
+
+    # Get all node IDs from any successful response
+    all_node_ids = set()
+    for checker in successful_checkers:
+        if checker.brokers_data:
```

Comment:
It only can evaluate to false if it's empty (and hence they all are). I assume it's not possible, as each node returns at least itself?